packages/sdk/wallets/wallet-keystore/src/keystores.ts

```diff
@@ -0,0 +1,148 @@
+import {
+  ensureLeading0x,
+  normalizeAddressWith0x,
+  privateKeyToAddress,
+} from '@celo/utils/lib/address'
+import Wallet from 'ethereumjs-wallet'
+import { mkdirSync, promises as fsPromises, readFileSync, unlinkSync, writeFileSync } from 'fs'
+import path from 'path'
+
+export enum ErrorMessages {
+  KEYSTORE_ENTRY_EXISTS = 'Existing encrypted keystore for address',
+  NO_MATCHING_ENTRY = 'Keystore entry not found for address',
+  UNKNOWN_STRUCTURE = 'Unexpected keystore entry structure',
+}
+
+export abstract class KeystoreBase {
+  /**
+   * Handles encrytion and transformation between private keys <-> V3Keystore strings
+   */
+
+  // TODO docstrings
+
+  // Must be implemented by subclass
+  // TODO could make all of these either keystoreName OR address
+  abstract persistKeystore(keystoreName: string, keystore: string): void
+  abstract getRawKeystore(keystoreName: string): string
+  abstract getAllKeystoreNames(): Promise<string[]>
+  abstract removeKeystore(keystoreName: string): void
+
+  getAddress(keystoreName: string): string {
+    const rawKeystore = this.getRawKeystore(keystoreName)
+    try {
+      const address = JSON.parse(rawKeystore).address
+      return ensureLeading0x(address)
+    } catch (e) {
+      console.log(e)
+      throw new Error(ErrorMessages.UNKNOWN_STRUCTURE)
+    }
+  }
+
+  async listKeystoreAddresses(): Promise<string[]> {
+    return Object.keys(await this.getAddressMap())
+  }
+
+  // Map addresses to their respective keystore entries (names)
+  async getAddressMap(): Promise<Record<string, string>> {
+    // Don't store this to minimize race conditions (file is deleted/added manually)
+    const addressToFile: Record<string, string> = {}
+    ;(await this.getAllKeystoreNames()).map((file) => (addressToFile[this.getAddress(file)] = file))
+    return addressToFile
+  }
+
+  async importPrivateKey(privateKey: string, passphrase: string) {
+    // Only allow for new private keys to be imported into the keystore
```

Comment:
Why? Is this behaviour consistent with Geth/Clef?

I'd expect idempotence here

Reply:
`new` is probably the wrong word to use in this comment (will fix/rephrase); better phrasing would be "only keys without existing entries can be imported". This is consistent with geth/clef (clef I think takes it a bit further and makes it really difficult to import pks as all but manages accounts that it creates), in that you can change the passphrase of an existing keystore or import a new one, but it freaks out if there are two keystores for the same account (i.e. if you manually move one in).

As far as idempotence, reasonable assumption but it turns out that because of how the scrypt key derivation function (KDF) works (I believe it's the salt that causes this, but my understanding here is shaky :sweat-smile: ), the same PK + password combo used multiple times creates different keystore files.